src/Evaluator.hs

```diff
@@ -38,8 +103,343 @@ data EvalError
   | EvalErrorType TypeError
   | EvalErrorCase SrcLoc Value
   | EvalErrorUser Syntax
-  deriving (Show)
+  | EvalErrorIdent Value
+  deriving (Show, Typeable)
 makePrisms ''EvalError
+instance Exception EvalError
+
+-- | The Kontinuation type. The naming convention InFoo means that the subject
+-- of evaluation in the CEK machine is Foo. For example, when the continuation
+-- is 'InArg' the subject of evaluation is the argument of the function and the
+-- continuation holds the un-evaluated function symbol in its first field.
+data Kont where
+
+  Halt :: Kont
+  -- ^ Marks the evaluator finishing
+
+  -- functions
+  InArg :: !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
+  -- ^ The argument is being evaluated, so hold onto the function symbol.
+  InFun :: !Value -> !VEnv -> !Kont -> Kont
+  -- ^ The function is being evaluated, so hold onto the evaluated argument.
+
+  InLetDef :: !Ident -> !Var -> !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
+  -- ^ Evaluating the let def
+
+  -- constructors
+  InCtor :: !Constructor -> ![CoreF TypePattern ConstructorPattern Core] -> ![Value] -> !VEnv -> !Kont -> Kont
```

Comment:
According to the naming convention, shouldn't it be `InCtorArg`? We're inside `arg_i` of `(Ctor arg_1 ... arg_i ... arg_n)`, not inside `Ctor`, right?

Reply:
And `arg_1` through `arg_{i-1}` have already been evaluated to `Value`s. Shouldn't the `[Value]` list be before the `[Core]` list? Seems weird to use the order `[arg_{i+1}, ..., arg_n] [arg_1, ..., arg_{i-1}]`. Or are both lists reversed? Is it `[arg_n, arg_{n-1}, ..., arg_{i+1}] [arg_{i-1}, arg_{i-2}, ..., arg_1]`?

Reply:
Ah indeed this is an inconsistency, I believe I was trying to keep the value list in registers instead of spilling it onto the stack. But I doubt this actually matters in practice.

Reply:
Wow! By all means, don't let me break your micro-optimizations, you are clearly having fun doing it! But when it is that obscure, please document them because otherwise I will most certainly break them by accident.